Change arg_kwarg_vals propagation strategy

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #150511
* #148104
* #150495
* #148092
* #148091
* #148063
* __->__ #148046

Instead of always propagating arg_kwarg_vals in _COPY_META_FIELDS, we
special-case the pattern matcher to propagate arg_kwarg_vals when
it sees triton_kernel_wrapper_functional.

The strategy is:
1) trace out the replacement graph with arg_kwarg_vals (which have accurate eager-mode metadata)
2) trace out the replacement graph with vals (which have the accurate Inductor metadata)
3) Propagate the arg_kwarg_vals from the first graph to the second.
4) Use the second graph as the replacement graph.

The strategy is this because we want to extend this to handle
auto_functionalized later up in the stack.

Test Plan:
- existing tests

cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

zou3519 committed:
Don't propagate arg_kwarg_vals

I changed the design to not propagate arg_kwarg_vals. Instead, a pass
author must manually figure out how to propagate this if it exists.
This seemed like the cleanest solution. In the future I will extend
arg_kwarg_vals to custom operators and the auto_functionalized HOP.

There's an alternative that we hard-code specific propagation rules into
the pattern matcher.
For example, triton_kernel_wrapper_functional may be allowed to propagate
this field to triton_kernel_wrapper_mutation. This seemed not ideal: in
the previous state, we were actually propagating
triton_kernel_wrapper_functional's arg_kwarg_vals field to the entire
replacement subgraph (includes triton_kernel_wrapper_mutation AND some clone nodes).
Furthermore, someone might create a pattern matcher pass that transforms
a single triton_kernel_wrapper_functional into multiple
triton_kernel_wrapper_mutation, and then the propagation is wrong.

For safety, the pattern matcher will error out by default if someone tries
to replace_by_example a node with arg_kwarg_vals (and therefore drop
that value).

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "Change arg_kwarg_vals propagation strategy"


Instead of always propagating arg_kwarg_vals in _COPY_META_FIELDS, we
teach the Inductor Pattern Matcher to trace out replacement graphs using
the arg_kwarg_vals *instead of* the node.meta['val'].

This makes more sense and fixes some subtle issues (before this PR, we
would propagate triton_kernel_wrapper_functional's arg_kwarg_vals to all
nodes in the decomposition, which includes
triton_kernel_wrapper_mutation and aten.clone nodes). This also works a
lot better if someone wants to replace a custom op with 1+ custom
ops (where all of them have arg_kwarg_vals).

Here's how it gets propagated.
- the triton_kernel_wrapper_functional has arg_kwarg_vals
- the pattern matcher traces out a replacement graph using
  the FakeTensors from arg_kwarg_vals
- this will end up using make_fx to trace a
  triton_kernel_wrapper_mutation into the graph
- when triton_kernel_wrapper_mutation gets traced, make_fx sets its
  arg_kwarg_vals field. This will be identical to the arg_kwarg_vals we
  used to trace the decomp.

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "Change arg_kwarg_vals propagation strategy"


Instead of always propagating arg_kwarg_vals in _COPY_META_FIELDS, we
teach the Inductor Pattern Matcher to trace out replacement graphs using
the arg_kwarg_vals *instead of* the node.meta['val'].

This makes more sense and fixes some subtle issues (before this PR, we
would propagate triton_kernel_wrapper_functional's arg_kwarg_vals to all
nodes in the decomposition, which includes
triton_kernel_wrapper_mutation and aten.clone nodes). This also works a
lot better if someone wants to replace a custom op with 1+ custom
ops (where all of them have arg_kwarg_vals).

Here's how it gets propagated.
- the triton_kernel_wrapper_functional has arg_kwarg_vals
- the pattern matcher traces out a replacement graph using
  the FakeTensors from arg_kwarg_vals
- this will end up using make_fx to trace a
  triton_kernel_wrapper_mutation into the graph
- when triton_kernel_wrapper_mutation gets traced, make_fx sets its
  arg_kwarg_vals field. This will be identical to the arg_kwarg_vals we
  used to trace the decomp.

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "Change arg_kwarg_vals propagation strategy"


Instead of always propagating arg_kwarg_vals in _COPY_META_FIELDS, we
teach the Inductor Pattern Matcher to trace out replacement graphs using
the arg_kwarg_vals *instead of* the node.meta['val'].

This makes more sense and fixes some subtle issues (before this PR, we
would propagate triton_kernel_wrapper_functional's arg_kwarg_vals to all
nodes in the decomposition, which includes
triton_kernel_wrapper_mutation and aten.clone nodes). This also works a
lot better if someone wants to replace a custom op with 1+ custom
ops (where all of them have arg_kwarg_vals).

Here's how it gets propagated.
- the triton_kernel_wrapper_functional has arg_kwarg_vals
- the pattern matcher traces out a replacement graph using
  the FakeTensors from arg_kwarg_vals
- this will end up using make_fx to trace a
  triton_kernel_wrapper_mutation into the graph
- when triton_kernel_wrapper_mutation gets traced, make_fx sets its
  arg_kwarg_vals field. This will be identical to the arg_kwarg_vals we
  used to trace the decomp.

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "Change arg_kwarg_vals propagation strategy"


Instead of always propagating arg_kwarg_vals in _COPY_META_FIELDS, we
teach the Inductor Pattern Matcher to trace out replacement graphs using
the arg_kwarg_vals *instead of* the node.meta['val'].

This makes more sense and fixes some subtle issues (before this PR, we
would propagate triton_kernel_wrapper_functional's arg_kwarg_vals to all
nodes in the decomposition, which includes
triton_kernel_wrapper_mutation and aten.clone nodes). This also works a
lot better if someone wants to replace a custom op with 1+ custom
ops (where all of them have arg_kwarg_vals).

Here's how it gets propagated.
- the triton_kernel_wrapper_functional has arg_kwarg_vals
- the pattern matcher traces out a replacement graph using
  the FakeTensors from arg_kwarg_vals
- this will end up using make_fx to trace a
  triton_kernel_wrapper_mutation into the graph
- when triton_kernel_wrapper_mutation gets traced, make_fx sets its
  arg_kwarg_vals field. This will be identical to the arg_kwarg_vals we
  used to trace the decomp.

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "Change arg_kwarg_vals propagation strategy"


Instead of always propagating arg_kwarg_vals in _COPY_META_FIELDS, we
teach the Inductor Pattern Matcher to trace out replacement graphs using
the arg_kwarg_vals *instead of* the node.meta['val'].

This makes more sense and fixes some subtle issues (before this PR, we
would propagate triton_kernel_wrapper_functional's arg_kwarg_vals to all
nodes in the decomposition, which includes
triton_kernel_wrapper_mutation and aten.clone nodes). This also works a
lot better if someone wants to replace a custom op with 1+ custom
ops (where all of them have arg_kwarg_vals).

Here's how it gets propagated.
- the triton_kernel_wrapper_functional has arg_kwarg_vals
- the pattern matcher traces out a replacement graph using
  the FakeTensors from arg_kwarg_vals
- this will end up using make_fx to trace a
  triton_kernel_wrapper_mutation into the graph
- when triton_kernel_wrapper_mutation gets traced, make_fx sets its
  arg_kwarg_vals field. This will be identical to the arg_kwarg_vals we
  used to trace the decomp.

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "Change arg_kwarg_vals propagation strategy"


Instead of always propagating arg_kwarg_vals in _COPY_META_FIELDS, we
teach the Inductor Pattern Matcher to trace out replacement graphs using
the arg_kwarg_vals *instead of* the node.meta['val'].

This makes more sense and fixes some subtle issues (before this PR, we
would propagate triton_kernel_wrapper_functional's arg_kwarg_vals to all
nodes in the decomposition, which includes
triton_kernel_wrapper_mutation and aten.clone nodes). This also works a
lot better if someone wants to replace a custom op with 1+ custom
ops (where all of them have arg_kwarg_vals).

Here's how it gets propagated.
- the triton_kernel_wrapper_functional has arg_kwarg_vals
- the pattern matcher traces out a replacement graph using
  the FakeTensors from arg_kwarg_vals
- this will end up using make_fx to trace a
  triton_kernel_wrapper_mutation into the graph
- when triton_kernel_wrapper_mutation gets traced, make_fx sets its
  arg_kwarg_vals field. This will be identical to the arg_kwarg_vals we
  used to trace the decomp.

Test Plan:
- existing tests

[ghstack-poisoned]

eellison commented on torch/_inductor/pattern_matcher.py:
```diff
@@ -259,9 +259,33 @@ def replace_by_example(
-            replacement = trace_fn(
-                replacement_fn, torch.fx.map_arg(args, lambda arg: arg.meta["val"])
-            )
+            if len(self.nodes) == 1 and "arg_kwarg_vals" in self.nodes[0].meta:
```
Hmm, i dont think we should use arg_kwarg_vals here. i think we should use whatever the most recent meta is. which is what the current case is.

zou3519 replied:
Fair enough. The reason I did it this way is because it makes dealing with auto_functionalized ra lot easier, but yeah I agree tracing with arg_kwarg_vals is not the right call here

eellison commented on torch/_inductor/pattern_matcher.py:
```diff
@@ -1087,6 +1111,8 @@ def run_node(self, node: torch.fx.Node) -> Any:
                         old_node=node,
                         pass_name="Interpreter_Replacer",
                     )
+                    if "arg_kwarg_vals" in node.meta:
```
Similarly, I'm wary of extending this too generally. If you replace one operator with a different operator, does it have the same generic input strides requirements ? I dont think we can say that generally. Also the schema could be different

zou3519 replied:
Let's say we're replacing an operator foo with a subgraph that includes the operator bar. This line isn't propagating foo's arg_kwarg_vals to bar's arg_kwarg_vals.

Instead, the sequence of events is:
- we trace out the subgraph. This subgraph's nodes may have arg_kwarg_vals in it (if proxy_tensor.py decided that those need to exist),
- we copy all nodes in subgraph to the graph the PatternMatcher is editing. When we copy the subgraph, we propagate certain meta vals over.  This line says that if the subgraph already has arg_kwarg_vals, then we should propagate it.

zou3519 committed:
Update on "Change arg_kwarg_vals propagation strategy"


Instead of always propagating arg_kwarg_vals in _COPY_META_FIELDS, we
special-case the pattern matcher to propagate arg_kwarg_vals when
it sees triton_kernel_wrapper_functional.

The strategy is:
1) trace out the replacement graph with arg_kwarg_vals (which have accurate eager-mode metadata)
2) trace out the replacement graph with vals (which have the accurate Inductor metadata)
3) Propagate the arg_kwarg_vals from the first graph to the second.
4) Use the second graph as the replacement graph.

The strategy is this because we want to extend this to handle
auto_functionalized later up in the stack.

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "Change arg_kwarg_vals propagation strategy"


Instead of always propagating arg_kwarg_vals in _COPY_META_FIELDS, we
special-case the pattern matcher to propagate arg_kwarg_vals when
it sees triton_kernel_wrapper_functional.

The strategy is:
1) trace out the replacement graph with arg_kwarg_vals (which have accurate eager-mode metadata)
2) trace out the replacement graph with vals (which have the accurate Inductor metadata)
3) Propagate the arg_kwarg_vals from the first graph to the second.
4) Use the second graph as the replacement graph.

The strategy is this because we want to extend this to handle
auto_functionalized later up in the stack.

Test Plan:
- existing tests

[ghstack-poisoned]

eellison commented on torch/_inductor/pattern_matcher.py:
```diff
@@ -251,14 +251,73 @@ def replace_by_example(
             else contextlib.nullcontext()
         )
 
+        def should_propagate_arg_kwarg_vals(nodes: list[torch.fx.Node]) -> bool:
```
is the a reason to not just have this take in a single node, if a list >= 1 will return False ?

zou3519 replied:
The check for >= 1 needs to exist somewhere, I put it into the function instead of outside of it. I can put it outside if you think that's clearer

eellison commented on torch/_inductor/pattern_matcher.py:
```diff
@@ -251,14 +251,73 @@ def replace_by_example(
+
+            if should_propagate_arg_kwarg_vals(self.nodes):
+                # Our strategy is:
+                # 1) trace out the graph with arg_kwarg_vals (which have accurate eager-mode metadata)
+                # 2) trace out the graph with vals (which have the accurate Inductor metadata)
```
We should be able to short-circuit this is they are equivalent metadata.

zou3519 replied:
What do you mean?

eellison commented on torch/_inductor/pattern_matcher.py:
```diff
@@ -251,14 +251,73 @@ def replace_by_example(
+                example_vals = torch.fx.map_arg(args, lambda arg: arg.meta["val"])
+                replacement = trace_fn(graph_with_eager_vals, example_vals)
+
+                # propagate metadata from first graph to second
+                # NB: This assertion might not be true in general, but it is true for
+                # the two use cases we have
+                # (triton_kernel_wrapper_functional, auto_functionalized)
+                assert len(graph_with_eager_vals.graph.nodes) == len(
+                    replacement.graph.nodes
+                )
```
Yes, this is not general... we can revisit when we extend.

eellison commented on torch/_inductor/pattern_matcher.py:
```diff
@@ -1083,6 +1142,11 @@ def run_node(self, node: torch.fx.Node) -> Any:
+                    # This function copy-pastes the replacement graph into
+                    # the graph. If the replacement graph had any arg_kwarg_vals,
+                    # or val/tensor_meta, we propagate those over.
+                    if "arg_kwarg_vals" in node.meta:
```
I"m a bit confused - isn't this what the above `should_propagate_arg_kwarg_vals` case is handling ?

zou3519 replied:
`replace_with_graph` has some logic where we take the replacement graph and copy it into the original graph. This step would otherwise lose the arg_kwarg_vals metadata.

eellison commented on torch/_inductor/pattern_matcher.py:
```diff
@@ -251,14 +251,73 @@ def replace_by_example(
+            node = nodes[0]
+            if "arg_kwarg_vals" not in node.meta:
+                return False
+            return node.target in OrderedSet(
```
nit: node.target is ...

zou3519 replied:
the code looks like this so I can add more things to it in the follow-up PR. I can change this here if you want me to but it's easier to not

eellison reviewed (commented):
Looks good - would it be possible to add tests ? (similar to https://github.com/pytorch/pytorch/blob/a7fe685be8b83c7bf0478c4b1dddf0c24f8d8790/test/inductor/test_triton_kernels.py#L3387-L3454).

It's also not clear to me we're handling ``replace_with_graph` with the same care as we are handling `replace_by_example`

zou3519 commented:
@eellison 

> Looks good - would it be possible to add tests ? (similar to

What kind of a test were you looking for? https://github.com/pytorch/pytorch/blob/a7fe685be8b83c7bf0478c4b1dddf0c24f8d8790/test/inductor/test_triton_kernels.py#L3387-L3454 (and all of the triton kernel reinplacing tests) are the test case for this.

> It's also not clear to me we're handling ``replace_with_graphwith the same care as we are handlingreplace_by_example`

I don't think replace_with_graph should have the same handling as replace_by_example. The idea being if you want to replace with a graph then I expect you are going to set the node.meta on said graph and we will not override them.

zou3519 commented:
We discussed this offline. The resolution was:
1) register_replacement should call replace_by_example (it does something similar). This would allow register_replacement to also get the "achieve args_kwargs_vals propagation" behavior. I'll do this as a follow-up.
2) tests - this PR just modifies the triton kernel path, not the custom op path, so the tests are sufficient. Later up in the stack I introduce custom op PRs.
3) not sure if you had any other comments/responses to the inline comments @eellison?

eellison reviewed (approved):
Yes, TODO: handle register_replacement. this is a more commonly used api than replace_by_example.

zou3519 committed:
Update on "Change arg_kwarg_vals propagation strategy"


Instead of always propagating arg_kwarg_vals in _COPY_META_FIELDS, we
special-case the pattern matcher to propagate arg_kwarg_vals when
it sees triton_kernel_wrapper_functional.

The strategy is:
1) trace out the replacement graph with arg_kwarg_vals (which have accurate eager-mode metadata)
2) trace out the replacement graph with vals (which have the accurate Inductor metadata)
3) Propagate the arg_kwarg_vals from the first graph to the second.
4) Use the second graph as the replacement graph.

The strategy is this because we want to extend this to handle
auto_functionalized later up in the stack.

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "Change arg_kwarg_vals propagation strategy"


Instead of always propagating arg_kwarg_vals in _COPY_META_FIELDS, we
special-case the pattern matcher to propagate arg_kwarg_vals when
it sees triton_kernel_wrapper_functional.

The strategy is:
1) trace out the replacement graph with arg_kwarg_vals (which have accurate eager-mode metadata)
2) trace out the replacement graph with vals (which have the accurate Inductor metadata)
3) Propagate the arg_kwarg_vals from the first graph to the second.
4) Use the second graph as the replacement graph.

The strategy is this because we want to extend this to handle
auto_functionalized later up in the stack.

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 commented:
> Yes, TODO: handle register_replacement. this is a more commonly used api than replace_by_example.

Thanks, I'll land this stack first and then finish the TODO. The top PR in the stack is risky so I want to see signals on that earlier

zou3519 committed:
Update on "Change arg_kwarg_vals propagation strategy"


Instead of always propagating arg_kwarg_vals in _COPY_META_FIELDS, we
special-case the pattern matcher to propagate arg_kwarg_vals when
it sees triton_kernel_wrapper_functional.

The strategy is:
1) trace out the replacement graph with arg_kwarg_vals (which have accurate eager-mode metadata)
2) trace out the replacement graph with vals (which have the accurate Inductor metadata)
3) Propagate the arg_kwarg_vals from the first graph to the second.
4) Use the second graph as the replacement graph.

The strategy is this because we want to extend this to handle
auto_functionalized later up in the stack.

Test Plan:
- existing tests

[ghstack-poisoned]
